system/trace/trace_dump.c

```diff
@@ -0,0 +1,643 @@
+           * Preserve the information for the succeeding NOTE_RESUME.
+           */
+
+          cctx->current_state = nsu->nsu_state;
+        }
+        break;
+
+      case NOTE_RESUME:
+        {
+          /* This note informs the task to be resumed.
+           * The task switch timing depends on the running context.
+           */
+
+          cctx->next_pid = pid;
+
+          if (!cctx->ininterrupt)
+            {
+              /* If not in the interrupt context, the task switch is
+               * executed immediately.
+               */
+
+              trace_dump_header(out, note, ctx);
+              trace_dump_sched_switch(out, note, ctx);
+            }
+          else
+            {
+              /* If in the interrupt context, the task switch is postponed
+               * until leaving the interrupt handler.
+               */
+
+              trace_dump_header(out, note, ctx);
+              fprintf(out, "sched_waking: comm=%s pid=%d target_cpu=%d\n",
+                      get_task_name(cctx->next_pid, ctx),
+                      get_pid(cctx->next_pid), cpu);
+              cctx->pendingswitch = true;
+            }
+        }
+        break;
+
+#ifdef CONFIG_SCHED_INSTRUMENTATION_SYSCALL
+      case NOTE_SYSCALL_ENTER:
+        {
+          FAR struct note_syscall_enter_s *nsc;
+          FAR struct trace_dump_task_context_s *tctx;
+
+          /* Exclude the case of syscall issued by an interrupt handler and
+           * nested syscalls to correct tracecompass display.
+           */
+
+          if (cctx->ininterrupt)
+            {
+              break;
+            }
+
+          tctx = get_task_context(pid, ctx);
+          if (tctx == NULL)
+            {
+              break;
+            }
+
+          tctx->syscall_nest++;
+          if (tctx->syscall_nest > 1)
+            {
+              break;
+            }
+
+          nsc = (FAR struct note_syscall_enter_s *)p;
+          if (nsc->nsc_nr < CONFIG_SYS_RESERVED ||
+              nsc->nsc_nr >= SYS_maxsyscall)
+            {
+              break;
+            }
+
+          trace_dump_header(out, note, ctx);
+          fprintf(out, "sys_%s()\n",
+                  g_funcnames[nsc->nsc_nr - CONFIG_SYS_RESERVED]);
+        }
+        break;
+
+      case NOTE_SYSCALL_LEAVE:
+        {
+          FAR struct note_syscall_leave_s *nsc;
+          FAR struct trace_dump_task_context_s *tctx;
+
+          /* Exclude the case of syscall issued by an interrupt handler and
+           * nested syscalls to correct tracecompass display.
+           */
+
+          if (cctx->ininterrupt)
+            {
+              break;
+            }
+
+          tctx = get_task_context(pid, ctx);
+          if (tctx == NULL)
+            {
+              break;
+            }
+
+          tctx->syscall_nest--;
+          if (tctx->syscall_nest > 0)
+            {
+              break;
+            }
+
+          tctx->syscall_nest = 0;
```

Comment:
It is needed because syscall enter and leave in the trace data are not always coupled.
If only syscall enter is overwritten and lost in the trace data, the first syscall leave may make nest to negative value.